Async batch

dincamihai opened the pull request:
### What does this PR do?

Implementation of [RFC#0002](https://github.com/saltstack/salt/blob/develop/rfcs/0002-netapi-async-batch.md)

### What issues does this PR fix or reference?

When doing a request to /run endpoint with the following payload:

```
{
    "client": "local_async",
    "eauth": "auto",
    "username": "admin",
    "password": "admin",
    "tgt": "*",
    "fun": "cmd.run",
    "arg": ["sleep $((RANDOM % 30)) && echo hello"],
    "timeout": 5,
    "gather_job_timeout": 5,
    "batch_delay": 1,
    "batch": "2"
}
```

A response is returned immediately. The response is empty (please ignore that for now) but it basically looks like this: https://github.com/saltstack/salt/blob/7f6e5d89a48bccf19cb45ed52a3a54f5cc53f400/salt/master.py#L2070-L2077

After returning the response, the batch executes a test.ping. Once the test.ping is done an event is published `salt/batch/<batch-jid>/start` with the following data: https://github.com/saltstack/salt/blob/7d4cae95f3fd54f10ad79544d4c9a2074ab11ed4/salt/cli/batch_async.py#L190-L194

When the batch finishes, it fires `salt/batch/<batch-jid>/done` event: https://github.com/saltstack/salt/blob/7d4cae95f3fd54f10ad79544d4c9a2074ab11ed4/salt/cli/batch_async.py#L199-L205

### Previous Behavior
The response is only returned after the batch job was executed.

### New Behavior
The response is returned immediately and the batch job continues to run as tornado coroutines.

### Tests written?

TODO

Commit:
Async batch

dincamihai commented:
@moio JFYI and, of course, ideas are welcome.

moio commented:
What is the value of `gather_job_timeout` used in [cmd_iter](https://github.com/saltstack/salt/blob/73d7089f64f31d90088a97057041aa1be3a687f7/salt/cli/batch.py#L60) in the two cases? Is it the same or different?

dincamihai commented:
@moio it is the same value, 10s.
Increasing to 30s for the API call does not result in more minions replying to the ping, it just waits longer for the minion that did not reply.
See below the effect of the 30s gather_job_timeout.
It waits from 10:39:29,861 to 10:39:59,969

```
2018-11-20 10:39:24,734 - master._do_batching - running
2018-11-20 10:39:24,735 - master._run_batches - running
2018-11-20 10:39:24,735 - master._run_batches - done
2018-11-20 10:39:24,757 - batch.__gather_minions - 30
2018-11-20 10:39:24,760 - master.publish - 20181120103924759543 test.ping * []
2018-11-20 10:39:24,816 - master._return - 20181120103924759543 id_JfQHm test.ping
2018-11-20 10:39:24,821 - master._return - 20181120103924759543 id_QRXlv test.ping
2018-11-20 10:39:24,834 - master._return - 20181120103924759543 id_afdeZ test.ping
2018-11-20 10:39:24,852 - master._return - 20181120103924759543 id_qyUbO test.ping
2018-11-20 10:39:29,837 - master.publish - 20181120103929834283 saltutil.find_job ['id_LPOqE'] ['20181120103924759543']
2018-11-20 10:39:29,861 - master._return - 20181120103929834283 id_LPOqE saltutil.find_job
2018-11-20 10:39:59,969 - master.publish - 20181120103959966032 cmd.run ['id_QRXlv', 'id_afdeZ'] ['sleep $((RANDOM % 3)) && echo hello']                                                                                                                                      
2018-11-20 10:40:01,248 - master._return - 20181120103959966032 id_QRXlv cmd.run
2018-11-20 10:40:02,578 - master.publish - 20181120104002577099 cmd.run ['id_JfQHm'] ['sleep $((RANDOM % 3)) && echo hello']
2018-11-20 10:40:02,860 - master._return - 20181120104002577099 id_JfQHm cmd.run
2018-11-20 10:40:03,632 - master.publish - 20181120104003630855 cmd.run ['id_qyUbO'] ['sleep $((RANDOM % 3)) && echo hello']
2018-11-20 10:40:05,678 - master.publish - 20181120104005674763 saltutil.find_job ['id_afdeZ'] ['20181120103959966032']
2018-11-20 10:40:05,723 - master._return - 20181120104005674763 id_afdeZ saltutil.find_job
2018-11-20 10:40:08,743 - master.publish - 20181120104008741501 saltutil.find_job ['id_qyUbO'] ['20181120104003630855']
2018-11-20 10:40:08,788 - master._return - 20181120104008741501 id_qyUbO saltutil.find_job
2018-11-20 10:40:38,926 - batch.run - done
2018-11-20 10:40:38,927 - master._return - 20181120103924759543 id_LPOqE test.ping
2018-11-20 10:40:38,933 - master._return - 20181120103959966032 id_afdeZ cmd.run
2018-11-20 10:40:38,936 - master._return - 20181120104003630855 id_qyUbO cmd.run
```

Something happens at some point and the test.ping returns are not coming anymore and then when the batching starts the batch job returns come and only after batch finishes the initial ping returns appear.

moio commented:
Do the pings reach the minions in the first place? What does `salt-run state.event pretty=True` say?

cachedout commented on salt/client/__init__.py:
```diff
@@ -1641,7 +1641,7 @@ def _prep_pub(self,
 
         # Generate the standard keyword args to feed to format_payload
-        payload_kwargs = {'cmd': 'publish',
+        payload_kwargs = {'cmd': kwargs.get('cmd', 'publish'),
```
I'm not following the logic behind this. Could you please clarify?

dincamihai replied:
from the RFC here https://github.com/saltstack/salt/blob/develop/rfcs/0002-netapi-async-batch.md

> a new method is added to ClearFuncs that is called in place of the publish for asynchronous batched command execution (eg. publish_batch). This method will receive batch-specific parameters and use publish internally

The change you are asking about was done in that direction, meaning that it allows me to specify the publish function I want to use from the master. This is unsafe at the moment since it allows calling any function on the master. This problem will be addressed before merging the PR.

dincamihai commented:
this PR includes now the refactoring from https://github.com/saltstack/salt/pull/49818

Commit:
Remove code used for debug

dincamihai commented:
@moio @cachedout I have updated my PR.
I have decided that the old implementation of Batch is not useful here so I have implemented the async version of batching almost from scratch.
It is using run_job_async to do the initial ping and to publish the job in batches. It is also registering an event handler to gather the returns.

Commit:
Fix `if` statement placement

cachedout commented:
@dincamihai Can we start with having you please fix these lint errors? https://jenkinsci.saltstack.com/job/pr-lint/job/PR-50546/13/warnings52Result/new/file.-1608698382/

Commit:
Remove unused imports

Commit:
Add find_job checks

Commit:
Check if should close on all events

dincamihai commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,174 @@
+# pylint: enable=import-error,no-name-in-module,redefined-builtin
+import logging
+import fnmatch
+
```
@cachedout do you know what this lint error means? https://jenkinsci.saltstack.com/job/pr-lint/job/PR-50546/28/warnings52Result/new/file.-1608698382/

cachedout replied:
```gray% pylint --rcfile=../../.testing.pylintrc batch_async.py                                                                                                                                                                (git)-[pr/50546]  
Using config file /Users/mikeplace/devel/salt/.testing.pylintrc                                                        
************* Module salt.cli.batch_async                  
batch_async.py:16: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:                    
 # pylint: enable=import-error,no-name-in-module,redefined-builtin                                                     
 import logging                                            
 import fnmatch                                            
+from salt.ext.six.moves import range                      
                                                           
 log = logging.getLogger(__name__)                         
                                                           

-----------------------------------                        
Your code has been rated at 9.89/10   ```

moio commented on salt/master.py:
```diff
@@ -2030,12 +2030,35 @@ def get_token(self, clear_load):
+        batch_load = {}
+        batch_load.update(clear_load)
+        batch_load['kwargs']['batch'] = batch
+        import salt.cli.batch_async
```
Is it intended to have import here and not at the beginning of the file?

moio replied:
:+1:

dincamihai replied:
keeping it

moio commented on salt/master.py:
```diff
@@ -2030,12 +2030,35 @@ def get_token(self, clear_load):
+        import salt.cli.batch_async
+        batch = salt.cli.batch_async.BatchAsync(
+            self.local.opts,
+            functools.partial(self._prep_jid, clear_load, {}),
```
Oh nice things you learn by doing PR reviews, `functools.partial`. Been looking for this for a while, thanks for sharing! :+1:

moio commented on salt/master.py:
```diff
@@ -2121,24 +2144,27 @@ def publish(self, clear_load):
-                'missing': missing
+        if batch:
+            return self.publish_batch(batch, clear_load)
+        else:
```
I guess the `else:` clause is not really necessary here since we `return`, and this would make the patch smaller/easier to see we are not changing the behaviour of this method unless the new parameter is set.

moio reviewed (commented):
Hey @dincamihai, this looks good!

I noted some questions/doubts inline. Please pardon my Python inexperience, some points might be trivial or downright wrong.

Apart from them and from tests which I understand being written, the main conceptual question is actually about the following note in the PR description:

> The response is empty (please ignore that for now) but it basically looks like this: {"return": [{}]}

What the caller is interested in, after the call is placed, is the list of minions expected to return (those that responded in time to the initial ping and that are thus included in the batch queue).

I can see two ways to address this:
 1. we make the initial ping part blocking, thus the response could contain a minion list (and still proceed asynchronously for the bulk of the batch)
 1. we return nothing/`true`/any other dummy value, and then provide a way to get to this list (I can presently only imagine via an event)

Frankly speaking, looking at current Uyuni needs, 1. would fit better. Still I will not hide the downside: any request could take up to `gather_job_timeout` seconds to return (`gather_job_timeout` seconds being the worst case that would happen when at least one targeted minion takes more than `gather_job_timeout` to respond, or is down).

How difficult is it to implement one or the other approach, or even even making it configurable?

Do you see aspects I am forgetting about in this discussion?

Do Salt maintainers have any specific remark?


Thanks again for all the efforts here!

dincamihai commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,174 @@
+'''
+
+# Import python libs
+from __future__ import absolute_import, print_function, unicode_literals
```
This is fine, the __future__ imports have a special meaning

from https://docs.python.org/2/reference/simple_stmts.html#future-statements:

```A future statement is recognized and treated specially at compile time: Changes to the semantics of core constructs are often implemented by generating different code. It may even be the case that a new feature introduces new incompatible syntax (such as a new reserved word), in which case the compiler may need to parse the module differently. Such decisions cannot be pushed off until runtime.```

moio replied:
Thanks for the clarification.

dincamihai commented on salt/cli/batch.py:
```diff
@@ -26,6 +26,79 @@
 log = logging.getLogger(__name__)
 
 
+def _get_bnum(opts, minions, quiet):
```
This is an artifact from the refactoring. I will remove the "_"

dincamihai commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,174 @@
+                        self.active.remove(minion)
+                        self.done_minions.add(minion)
+                        # call later so that we maybe gather more returns
+                        self.event.io_loop.call_later(1, self.next)
```
Yes, that's correct. I will make this a configurable parameter.

dincamihai replied:
`batch_delay` can now be used in the request as a parameter. I have updated the description of PR

dincamihai commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,174 @@
+                    self.initialized = True
+
+        if self.initialized and self.done_minions == self.minions.difference(self.timedout_minions):
+            # TODO
```
@moio i will remove the TODO since it is DONE.

Commit:
Remove TODO comment

Commit:
Fix pylint error

Commit:
Remove leading '_' from function names

dincamihai commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,174 @@
+            (find_job_return_pattern, 'find_job_return')
+        }
+        if not self.event.subscriber.connected():
+            # TODO is there a way to subscribe to only some tags?
```
@moio yes, but the implementation is like a filtering after the event handler is called. What I had in mind, but I could not find a way to do it was to register the event handler only for some patterns so that the filtering would happen before the event handler would even be called.

moio replied:
I would say, unless some upstream reviewer can indicate a way, you could just remove this TODO then. Thanks for the explanation.

dincamihai commented on salt/client/__init__.py:
```diff
@@ -1761,7 +1730,8 @@ def pub(self,
             payload = channel.send(payload_kwargs, timeout=timeout)
-        except SaltReqTimeoutError:
+        except SaltReqTimeoutError as err:
+            log.error(err)
```
@moio this is intended. That error log entry proved itself very useful while trying to understand why it is timing out. I think it would be a good idea to have it in the logs and not hide it with the general error message that follows.

moio replied:
:+1:

Commit:
Remove TODO comment

Commit:
Make batch_delay a request parameter

Commit:
Improvements

 - start batching immediatelly after all minions reply to ping
 - return jid, available minions and missing minions in the response
 - fire event when batching start and when it ends

Commit:
Allow multiple event handlers

Commit:
Use config value for gather_job_timeout when not in payload

Commit:
Add async batch unittests

Commit:
Remove unnecessary else

Commit:
Set utf8 and remove duplicate dict value

Commit:
Replace str substitution with format

dincamihai commented on salt/cli/batch_async.py:
```diff
@@ -0,0 +1,174 @@
+                jid=self.batch_jid,
+                **self.eauth)
+            # TODO add parameter for find_job - should use gather_job_timeout?
+            self.event.io_loop.call_later(10, self.find_job, set(next_batch))
```
Yes, that is the idea.

## Short story
From my tests it resulted that the `timeout` parameter is used for firing the `find_job`.
This payload:

```
{
    "client": "local_batch",
    "eauth": "auto",
    "username": "admin",
    "password": "admin",
    "tgt": "*",
    "fun": "cmd.run",
    "arg": ["sleep $((RANDOM % 3)) && echo hello"],
    "timeout": 5,
    "gather_job_timeout": 10,
    "batch": "2"
}
```
Had this result: (5 secs == timeout param == this long took for find_job ['A'] to fire)
```
2019-01-18 16:11:14,801 - master.publish - 20190118161114798800 test.ping * []
2019-01-18 16:11:14,935 - master._return - 20190118161114798800 C test.ping
2019-01-18 16:11:14,947 - master._return - 20190118161114798800 D test.ping
2019-01-18 16:11:14,953 - master._return - 20190118161114798800 E test.ping
2019-01-18 16:11:14,962 - master._return - 20190118161114798800 B test.ping
2019-01-18 16:11:19,841 - __init__.gather_job_info - 20190118161114798800 ['A']
2019-01-18 16:11:19,887 - master.publish - 20190118161119886050 saltutil.find_job ['A'] ['20190118161114798800']
2019-01-18 16:11:19,964 - master._return - 20190118161119886050 A saltutil.find_job
```

## Long story

If I understand correctly, the original `--batch` mode goes through this path: https://github.com/saltstack/salt/blob/6f760b7174efb9d41de4a6194190f4ed6122e564/salt/cli/batch.py#L234-L241
 ==> https://github.com/saltstack/salt/blob/6f760b7174efb9d41de4a6194190f4ed6122e564/salt/client/__init__.py#L906-L912 ==> https://github.com/saltstack/salt/blob/6f760b7174efb9d41de4a6194190f4ed6122e564/salt/client/__init__.py#L1152 ==> https://github.com/saltstack/salt/blob/6f760b7174efb9d41de4a6194190f4ed6122e564/salt/client/__init__.py#L229-L248

it seem the `gather_job_info` is triggered based on `timeout_at` https://github.com/saltstack/salt/blob/6f760b7174efb9d41de4a6194190f4ed6122e564/salt/client/__init__.py#L1150-L1152

and `timeout_at` is computed both using `timeout` https://github.com/saltstack/salt/blob/6f760b7174efb9d41de4a6194190f4ed6122e564/salt/client/__init__.py#L1086 and `gather_job_timeout` https://github.com/saltstack/salt/blob/6f760b7174efb9d41de4a6194190f4ed6122e564/salt/client/__init__.py#L1160

Commit:
Improve docstring

Commit:
Use  for subsequent

cachedout commented:
IMHO this is a really nice feature and it would be good to get this in. Maybe @thatch45 can take a moment and review it and perhaps we can move forward here?

Commit:
Rename  method

Commit:
Remove  import

Commit:
Fix pylint error

isbm commented:
> IMHO this is a really nice feature and it would be good to get this in. Maybe @thatch45 can take a moment and review it and perhaps we can move forward here?

WIP

Commit:
Fix tests - method rename

Commit:
Improve BatchAsync docstring

Commit:
Allow metadata to pass

Commit:
Pass metadata only to batch jobs

Commit:
Add the metadata to the start/done events

Commit:
Pass only metadata not all **kwargs

dincamihai commented:
@Ch3LL I would really appreciate if you could help me get this merged? :pray:

thatch45 commented:
@dincamihai sorry we have not looked over this for a while, I was just watching the commits come in and I was not sure you were done yet. Let me take a look :)

dincamihai commented:
@DmitryKuzmenko could you please take a look at this PR?

thatch45 commented:
Just to cover my bases, I am good with this PR. @DmitryKuzmenko won't be back until Monday, and I would still like him to take a look. But tests are passing and I think this is good to merge.

DmitryKuzmenko reviewed (approved):
I like this async approach! :+1:
